### DISCUSSION TITLE
Strange Polymorphic Mapper Error ... "not a sub-mapper of"...

### DISCUSSION THREAD

**Discussion by JonnyWaffles** [sqlalchemy/sqlalchemy#8509] [Usage Questions] [answered]
Hi team,

I'm trying to create a subclass of a third party package's model, specifically my own customized Airflow [BaseJob](https://github.com/apache/airflow/blob/3c9c0f940b67c25285259541478ebb413b94a73a/airflow/jobs/base_job.py#L50)

I expected to be able to easily subclass it and set a new `polymorphic_identity`.
```python
class SqsTriggererJob(BaseJob):
    __mapper_args__ = {"polymorphic_identity": "SqsTriggererJob"}
```

However, when I try to load the object from the ORM, I encounter the error message defined here https://github.com/sqlalchemy/sqlalchemy/blob/479dbc99e7fc5a60f846992c0cca8542047a8933/lib/sqlalchemy/orm/loading.py#L1350-L1361
```
sqlalchemy.exc.InvalidRequestError: Row with identity key (<class 'airflow.jobs.base_job.BaseJob'>, (13,), None) can't be loaded into an object; the polymorphic discriminator column 'job.job_type' refers to mapped class SqsTriggererJob->job, which is not a sub-mapper of the requested mapped class SqsTriggererJob->job
```
Strange right? So I fire up the Pycharm debugger and realize my issue has something to do with this block, specifically the last line:

https://github.com/sqlalchemy/sqlalchemy/blob/479dbc99e7fc5a60f846992c0cca8542047a8933/lib/sqlalchemy/orm/loading.py#L1313-L1324
As of line 1324 I can see this following values in the debugger
```
mapper = {Mapper} mapped class SqsTriggererJob->job
sub_mapper = {Mapper} mapped class SqsTriggererJob->job
```
When I try interactive debugging
```
>>> sub_mapper is mapper
False
```
Okay so they aren't the same object. Which is strange because they look identical. Both of their `inherits` properties point to the same object
```python
>>> sub_mapper.inherits
<Mapper at 0x7f17ed183f10; BaseJob>
>>> mapper.inherits
<Mapper at 0x7f17ed183f10; BaseJob>
```
Unfortunately this means the `isa` check fails too (after all they are identical!).

I don't know enough about how the ORM works to understand how this is occurring. I poked around Airflow's code looking for ORM side effects but I don't see any. My suspicion is this has something to do with model load ordering, ie all of Airflow's `BaseJob` models are loaded at once, and then I load mine later via plugin.

1. How is it possible that two identical looking mappers are being created, but are not the same object thus the `if sub_mapper is mapper` check is failing?
2. Any ideas for how to resolve this strangeness or where I may find the culprit?

**Comment by CaselIT** [comment]
Hi,

A runnable example that reproduce the issue would help debugging this on this side

**Reply by JonnyWaffles** [reply]
It's really hard to recreate outside the context of Airflow. Let me see if I can use @zzzeek 's tip below and I will follow up with a better working example if needed.

**Reply by CaselIT** [reply]
while not optimal, it would be fine also in the airflow context. Many times trying to create a runnable example helps finding the issue.

I agree that that following up mike suggestion is the best first step

**Comment by zzzeek** [comment, marked as answer]
Seems like the class is being defined twice.   So yes I'd look into this "plugin" thing seems problematic

**Reply by JonnyWaffles** [reply]
Oh I think you are on to something! Airflow does wonky stuff where it tries to create modules inside its plugin system. Will investigate and get back to you.

**Reply by JonnyWaffles** [reply]
Hooray thanks @zzzeek! As always you have saved me once again. I moved my code out of my Airflow plugins folder and made it a regular Python module and the issue was resolved. Something about the Airflow plugin machinery caused the above. Cheers and thanks again. How can I buy you a coffee(s)!?

**Reply by zzzeek** [reply]
shrugs, i never go anywhere :)   glad it working that's all that matters
